web/src/message_fetch.ts

```diff
@@ -715,6 +715,39 @@ export function initialize(finished_initial_fetch: () => void): void {
         } catch (error) {
             blueslip.error("Error in recent_view_ui.process_messages", undefined, error);
         }
+
+        // Since we backfill a lot more messages here compared to rendered message list,
+        // we can try populating them if we can do so locally.
+        for (const msg_list of message_lists.all_rendered_message_lists()) {
+            if (!msg_list.data.filter.can_apply_locally()) {
+                continue;
+            }
+
+            if (msg_list.empty() && msg_list.data.fetch_status.is_loading_messages()) {
+                // This is the first message fetch for this message list, we don't
+                // want to add messages right now, since we don't have a range of
+                // messages that user wants to see.
+                continue;
+            }
+
+            if (msg_list.visibly_empty()) {
+                // If the message list is visibly empty, we don't want to add
+                // message here to break the continuous message history.
+                // We will rely here on our backfill logic to render any visible
+                // messages if we can.
+                continue;
+            }
+
+            // This callback is only called when backfilling messages,
+            // so we need to check for the presence of any message from
+            // the message list in the all_messages_data to
+            // check for continuous message history for the message list.
+            const first_message = msg_list.first();
+            assert(first_message !== undefined);
+            if (all_messages_data.get(first_message.id) !== undefined) {
+                msg_list.add_messages(messages, {});
+            }
```

Comment:
I don't understand this last check; let me try to talk through it. I guess the thought is that if your `msg_list.first` is within the `all_messages_data` range, then we know that their contiguous windows overlap on that end, and so we can correctly extend this message feed with all messages from `all_messages_data` matching the filter. 

Maybe worth adjusting the comment to spell that out.

Oh, but there's a corner case that I'm not sure is correct. What if the view didn't originally populate from `all_messages_data`, because its anchor message wasn't in that window, but a later extension of the message feed for the target view has resulted in it now overlapping `all_messages_data`? Seems uncommon, but maybe possible. And there's another possibility where we might have previously skipped due to the `fetch_status.is_loading_messages` check above.

I think it might be then correct to try adding all of `all_messages_data` to the target list? `valid_non_duplicated_messages` should filter them more or less instantly... but it would still be quadratic work if `all_messages_data` grows to many 10Ks, and that's very bad. 

Oh but we should be able to just process the prefix of `all_messages_data` that's a lower message ID than `first_message`; shouldn't be too hard to compute that quickly.

Does that sound good?

Reply:
Yeah, did that. Sorry, I didn't realize I was just adding `messages` for only this fetch.